### DISCUSSION TITLE
Pipeline updates

### DISCUSSION THREAD

**Discussion by sgoldman101** [ld32/smarterSlurm#32] [Q&A]
If a single step in a pipeline is updated, for example changing the mapping step from BWA to Bowtie2, will the jobRecords for the unchanged steps be retained?

If the max number of jobRecords is capped, would there need to be a cleanup process for records belonging to steps that are no longer used?

**Comment by ld32** [comment]
Q: 
If a single step in a pipeline is updated, for example changing the mapping step from BWA to Bowtie2, will the jobRecords for the unchanged steps be retained?

A:
Yes, all old records are still retained with the original software name. 

By the way, if user changes the software commands, please remember to change the software name in the step decorator. For example from BWA to bowtie: 
#@3,1,bwa,i,,input,sbatch -p short -c 4 --mem 32G -t 2:0:0
to 
#@3,1,bowte,i,,input,sbatch -p short -c 4 --mem 32G -t 2:0:0

Otherwise, if user only modifies the software commands, but not the software name in the step decorator, new jobs will the wrong resource statistics to estimate mem/time to use, and the final actual mem/time usage still use the wrong software name. This can cause big problem, because the two sfotware likely have different resource consumption profiles.

Q:
If the max number of jobRecords is capped, would there need to be a cleanup process for records belonging to steps that are no longer used?

A:
Right now the cap is 200 records per software name. If it is full, new records still can still be added in if the input size is larger than all the earlier records. We can clean up the records to make sure the input sizes are more evenly distributed. If we are sure a job step (softwrae name) will not be run again, we can remove all the related records. We can also re-name the software name if we want to.
